contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,384 @@
+    }
+
+    /* ============ External Functions ============ */
+
+    receive() external payable {}
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a set token. This function must only be called before the first time
+     * this smart contract is used on any particular set token, or when a new token is added
+     * to the set token.
+     *
+     * @param _setToken    Address of the set token being initialized
+     */
+    function initApprovals(ISetToken _setToken) external {
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i=0; i<positions.length; i++) {
+            IERC20 token = IERC20(positions[i].component);
+            token.approve(address(uniRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+            token.approve(address(basicIssuanceModule), PreciseUnitMath.maxUint256());
+        }
+    }
+
+    /**
+     * Redeems a set token and sells the underlying tokens using Uniswap
+     * or Sushiswap.
+     *
+     * @param _setToken             Address of the set token being redeemed
+     * @param _amountSetToRedeem    The amount of the set token to redeem
+     * @param _isOutputETH          Set to true if the output token is Ether
+     * @param _outputToken          Address of output token. Ignored if _isOutputETH is true
+     * @param _minOutputReceive     Minimum amount of output token / ether to receive
+     */
+    function exchangeRedeem(ISetToken _setToken, uint256 _amountSetToRedeem, bool _isOutputETH, address _outputToken, uint256 _minOutputReceive) external nonReentrant {
+        _setToken.transferFrom(msg.sender, address(this), _amountSetToRedeem);
+        basicIssuanceModule.redeem(_setToken, _amountSetToRedeem, address(this));
+        liquidateComponents(_setToken);
+        if(_isOutputETH) {
+            IWETH(WETH).withdraw(IERC20(WETH).balanceOf(address(this)));
+            require(address(this).balance > _minOutputReceive, "INSUFFICIENT_OUTPUT_AMOUNT");
+            msg.sender.transfer(address(this).balance);
+        } else if (_outputToken == WETH) {
+            require(IERC20(WETH).balanceOf(address(this)) > _minOutputReceive, "INSUFFICIENT_OUTPUT_AMOUNT");
+            IERC20(WETH).transfer(msg.sender, IERC20(WETH).balanceOf(address(this)));
+        } else {
+            uint256 outputAmount = purchaseTokenBestPrice(_outputToken, IERC20(WETH).balanceOf(address(this)));
+            require(outputAmount > _minOutputReceive, "INSUFFICIENT_OUTPUT_AMOUNT");
+            IERC20(_outputToken).transfer(msg.sender, outputAmount);
+        }
+        emit ExchangeRedeem(msg.sender, address(_setToken), _isOutputETH ? address(0) : _outputToken, _amountSetToRedeem);
+    }
+
+    /**
+     * Gets the approximate amount of ETH or ERC20 gained when redeeming using
```

Comment:
Improve comment: `Returns an estimated quantity of ETH or specified ERC20 received for a given SetToken and SetToken quantity. Estimation pulls the best price of each component from Uniswap or Sushiswap.`